src/Microsoft.AspNet.Mvc.Xml/RequiredValidationHelper.cs

```diff
@@ -0,0 +1,209 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Reflection;
+using System.Runtime.Serialization;
+using Microsoft.AspNet.Mvc.ModelBinding;
+
+namespace Microsoft.AspNet.Mvc.Xml
+{
+    /// <summary>
+    /// Validates types having value type properties decorated with <see cref="RequiredAttribute"/>
+    /// but no <see cref="DataMemberAttribute"/>.
+    /// <see cref="JsonInputFormatter"/> supports <see cref="RequiredAttribute"/> where as the xml formatters
+    /// do not. Since a user's aplication can have both Json and Xml formatters, a request could be validated
+    /// when posted as Json but not Xml. So to prevent end users from having a false sense of security when posting
+    /// as Xml, we add errors to model-state to at least let the users know that there is a problem with their models.
+    /// </summary>
+    public class DataAnnotationRequiredAttributeValidation
+    {
+        private ConcurrentDictionary<Type, Dictionary<Type, List<string>>> _cachedValidationErrors
```

Comment:
why is this dictionary important? Is it because Input formatters are singleton generally so multiple concurrent access should be cached?